Code and Tests update

dipanjanS opened the pull request:
- Updated source code f068b0d
 - updated tests fcaa87a
- added tools scripts 4ea7c95
- Added requirements manifest file 9aef07c
- Added test execution script 229c634

dipanjanS committed:
Updated source code

dipanjanS committed:
updated tests

dipanjanS committed:
added tools scripts

dipanjanS committed:
Added requirements manifest file

dipanjanS committed:
Added test execution script

samuzzal-choudhury commented on runtest.sh:
```diff
@@ -0,0 +1,88 @@
+set -x
+
+# test coverage threshold
+COVERAGE_THRESHOLD=15
```
@dipanjanS I think that coverage threshold of 15 is too low. This has to be increased to 70-80. Please run this script locally to see the current code coverage number.

dipanjanS replied:
Hey @samuzzal-choudhury actually had a discussion with @tisnik and since we have to mock everything in the tests including BigQuery and S3 storage, the actual functions mostly are not being used in the tests, since otherwise it would involve calling BQ and S3 actually + they are from rudra. So he told me then lower coverage is ok, let me know what you think. Also main function would be the actual cron job which calls the other code but that usually doesn't show in coverage. We can discuss more on this also later as needed.

samuzzal-choudhury commented on runtest.sh:
```diff
@@ -0,0 +1,88 @@
+DISABLE_AUTHENTICATION=1 PYTHONDONTWRITEBYTECODE=1 python "$(which pytest)" --cov=./src/ --cov-report term-missing --cov-fail-under=$COVERAGE_THRESHOLD -vv ./tests/
+printf "%stests passed%s\n\n" "${GREEN}" "${NORMAL}"
+
+#`which codecov` --token=b2147eb8-2c22-486c-a417-220b36ceceb5
```
Please work with @tisnik to get a codecov token to be used with this repo.

dipanjanS replied:
Sure, @tisnik let me know if you can help me with this one

tisnik replied:
@dipanjanS use the following token please 5c0f0ca6-c3aa-407f-9b61-07830d9325e5

samuzzal-choudhury reviewed (commented):
@dipanjanS please see if the code conforms to the pylint and docstyle by running the scripts manually.

dipanjanS committed:
Added dockerfile

dipanjanS committed:
Added dockerfile for RHEL

dipanjanS committed:
Added openshift deployment template

dipanjanS committed:
Added code coverage, linting check scripts

dipanjanS committed:
Added setup scripts

dipanjanS committed:
updated and refactored source code

dipanjanS committed:
added requirements file

dipanjanS committed:
updated unit tests

dipanjanS committed:
added python version check script

dipanjanS commented:
@samuzzal-choudhury \ @tisnik updated all the aspects mentioned during the previous review. PTAL

samuzzal-choudhury commented on Dockerfile:
```diff
@@ -0,0 +1,24 @@
+WORKDIR ${APP_DIR}
+
+RUN pip3 install --upgrade pip
+RUN pip3 install git+https://github.com/fabric8-analytics/fabric8-analytics-rudra#egg=rudra
```
@dipanjanS are you planning to install a particular commit of the Rudra library or the latest master? Using a particular commit hash at times helps in preventing code crashes because of some changes in the library.

dipanjanS replied:
The latest version should be fine @samuzzal-choudhury like the recommendation models use, but if you think its better then we can switch to a specific commit version?

samuzzal-choudhury commented on Dockerfile.rhel:
```diff
@@ -0,0 +1,21 @@
+WORKDIR ${APP_DIR}
+
+RUN pip3 install --upgrade pip
+RUN pip3 install git+https://github.com/fabric8-analytics/fabric8-analytics-rudra#egg=rudra
```
@dipanjanS are you planning to install a particular commit of the Rudra library or the latest master? Using a particular commit hash at times helps in preventing code crashes because of some changes in the library.

dipanjanS replied:
The latest version should be fine @samuzzal-choudhury like the recommendation models use, but if you think its better then we can switch to a specific commit version?

samuzzal-choudhury commented on check-docstyle.sh:
```diff
@@ -0,0 +1,78 @@
+#!/bin/bash
+
+directories="src tests tools"
```
@dipanjanS there is no 'tools' directory in the repo. Please remove this from the directories' list.

dipanjanS replied:
@samuzzal-choudhury there is a tools directory but it is added in this commit itself: https://github.com/fabric8-analytics/fabric8-analytics-bigquery-manifests-job/tree/adbc22f8d423a09fdb1340924821881d6feab431/tools

that should work right? (same for all the similar comments below)

samuzzal-choudhury replied:
@dipanjanS  Please keep 'tools' in the directories' list of all the scripts if you are adding that directory in the repository. This comment goes for all other places where I mentioned the above comment.

dipanjanS replied:
Hey @samuzzal-choudhury sure, I think that is what I have done, followed the directory structure based on `https://github.com/fabric8-analytics/f8a-stacks-report`

samuzzal-choudhury commented on detect-common-errors.sh:
```diff
@@ -0,0 +1,76 @@
+#!/bin/bash
+
+directories="src tools tests"
```
@dipanjanS there is no 'tools' directory in the repo. Please remove this from the directories' list.

dipanjanS replied:
Same as above (added in this commit)

samuzzal-choudhury commented on detect-dead-code.sh:
```diff
@@ -0,0 +1,78 @@
+#!/bin/bash
+
+directories="src tools"
```
@dipanjanS there is no 'tools' directory in the repo. Please remove this from the directories' list.

dipanjanS replied:
Same as above (added in this commit)

samuzzal-choudhury commented on openshift/template.yaml:
```diff
@@ -0,0 +1,83 @@
+    displayName: Schedule
+    required: true
+    name: CRON_SCHEDULE
+    value: "*/10 * * * *"
```
@dipanjanS this join is set to run this joib every 10 minutes as per this CRON_SCHEDULE. Please change this to the appropriate schedule.

dipanjanS replied:
@samuzzal-choudhury what should be the ideal schedule do you think, can you send me the expression, i'll update this then

samuzzal-choudhury replied:
@dipanjanS it can be run monthly once. The expression is going to be '0 0 1 */1 *'. This means that the job runs at 00:00 hrs on 1st day of every month.

dipanjanS replied:
Thanks @samuzzal-choudhury just pushed it as a new commit now

samuzzal-choudhury commented on openshift/template.yaml:
```diff
@@ -0,0 +1,83 @@
+    displayName: Memory limit
+    required: true
+    name: MEMORY_LIMIT
+    value: "1024Mi" 
```
@dipanjanS please see that the CPU and Memory request and limit values are set as per the inferred values during the spike.

dipanjanS replied:
@samuzzal-choudhury sure I will verify this once more

samuzzal-choudhury commented on run-linter.sh:
```diff
@@ -0,0 +1,80 @@
+#!/bin/bash
+
+directories="src tests tools"
```
@dipanjanS there is no 'tools' directory in the repo. Please remove this from the directories' list.

dipanjanS replied:
Same as above (added in this commit itself)

samuzzal-choudhury commented on requirements.txt:
```diff
@@ -0,0 +1,5 @@
+codecov==2.0.15
```
@dipanjanS the requirements.txt file has to be created by using `pip_compile` command on a manually created `requirements.in`. You can rename this file to `requirements.in` and then run `pip_compile` to auto-generate a `requirements.txt`.

samuzzal-choudhury replied:
@dipanjanS looks good now.

samuzzal-choudhury commented on runtest.sh:
```diff
@@ -0,0 +1,88 @@
+DISABLE_AUTHENTICATION=1 PYTHONDONTWRITEBYTECODE=1 python "$(which pytest)" --cov=./src/ --cov-report term-missing --cov-fail-under=$COVERAGE_THRESHOLD -vv ./tests/
+printf "%stests passed%s\n\n" "${GREEN}" "${NORMAL}"
+
+`which codecov` --token=5c0f0ca6-c3aa-407f-9b61-07830d9325e5
```
@dipanjanS I guess the token was created by Pavel T for this repository. If not, please get in touch with him to have a proper condecov token.

dipanjanS replied:
Yes @tisnik has given this token @samuzzal-choudhury so we can use this

samuzzal-choudhury commented on src/config/cloud_constants.py:
```diff
@@ -7,7 +7,9 @@
 AWS_S3_ACCESS_KEY_ID = os.environ.get('AWS_S3_ACCESS_KEY_ID', '')
 AWS_S3_SECRET_ACCESS_KEY = os.environ.get('AWS_S3_SECRET_ACCESS_KEY', '')
-S3_BUCKET_NAME = 'developer-analytics-audit-report'
+S3_BUCKET_NAME = 'dsarkar-developer-analytics-audit-report'
```
Please change the S3 bucket name should be `developer-analytics-audit-report` for staging and production environments.

dipanjanS replied:
Sure @samuzzal-choudhury I am hoping once this job runs there it should be able to dump the data there? (is the bucket present already in staging? else it might fail)

samuzzal-choudhury commented on tests/test_manifests_job.py:
```diff
@@ -48,122 +58,148 @@ def manifest_content(self, eco):
 def _data_process_client_maven(_mock_bigquery_obj):
-    from rudra.data_store.bigquery.maven_bigquery import MavenBigQuery, MavenBQDataProcessing
-    _mvn_ins = MavenBigQuery()
+    _mvn_ins = manifests_job.mvn_bq.MavenBigQuery()
```
The docstyle run will fail without the docstrings for the test functions. This comment holds good for any function without a docstring.

dipanjanS replied:
I think it didn't throw up an error when I ran the check docstyle script, because maybe it's a fixture @samuzzal-choudhury ? But I will add in docstrings to this.

samuzzal-choudhury commented on tools/check_python_version.py:
```diff
@@ -0,0 +1,71 @@
+    if len(arguments) <= 2:
+        print("Usage python check_python_version.py major minor\n"
+              "      python3 check_python_version.py major minor")
+        raise Exception("CLI arguments missing")
```
The exceptions handled and raised in python has to be specific rather than the top level class `Exception`. This comment goes for all the occurrences in this code.

dipanjanS replied:
@samuzzal-choudhury actually I just reused this from stacks-report here: https://github.com/fabric8-analytics/f8a-stacks-report/blob/master/tools/check_python_version.py let me know if that works or should I change the exceptions to more specific ones?

tisnik replied:
@samuzzal-choudhury @dipanjanS do we need it for the simple script isolated from other code and called from CLI? I agree it is a good approach for larger code (and yet we still use it in our repositories ...)

dipanjanS replied:
For this case I think its just being used to check the python version and not used with the main codebase, so I'm good with it if you folks are :)

samuzzal-choudhury reviewed (changes requested):
@dipanjanS There are a few comments. PTAL.

dipanjanS committed:
added requirements.in file

dipanjanS committed:
added missing docstrings

dipanjanS committed:
updated s3 bucket name

dipanjanS committed:
updated dockerfiles

dipanjanS committed:
generated new requirements.txt file

dipanjanS committed:
updated scripts

dipanjanS commented:
@samuzzal-choudhury made the necessary changes as you suggested can you PTAL 
- auto-generated `requirements.txt`
- added specific commit hash to `rudra`
- added missing docstrings
- `tools` folder present in the commit itself
- codecov token present and supplied by @tisnik 
- updated bucket name based on your inputs

dipanjanS committed:
updated cron schedule

samuzzal-choudhury commented on openshift/template.yaml:
```diff
@@ -0,0 +1,83 @@
+            restartPolicy: Never
+            containers:
+            - name: bigquery-manifests
+              image: "djsarkar/bigquery-manifests:latest"
```
@dipanjanS this is not the right image. Please provide the quay.io rhel image reference as that is going to be deployed.

dipanjanS replied:
Done @samuzzal-choudhury

dipanjanS committed:
made s3 region configurable based on deployment template

dipanjanS committed:
updated deployment template with correct image name and other configs

dipanjanS commented:
@samuzzal-choudhury updated the image name and cron schedule and other deployment secrets as needed in the template. PTAL and then if its alright we can merge this.

Then I'll send another PR with the cico scripts.

samuzzal-choudhury reviewed (approved):
@dipanjanS Looks good. Please merge this PR.
